File: src/main/java/com/learning/java/controlstruct/ExceptionSyntax.java
package com.learning.java.controlstruct;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;

/* Java: Исключения (Exception) */


public class ExceptionSyntax {

	public static void main(String[] args) {
		/* try-catch-finally */
		try {
			URL url = new URL( "http://tirthalpatel.blogspot.com" );
			BufferedReader buffered = new BufferedReader( new InputStreamReader(url.openStream()) );
			try {
				String line;
				while((line = buffered.readLine()) != null)
					System.out.println( line );
			} finally {
                buffered.close();
			}
		} catch(MalformedURLException e) {  // ( ловим специальное исключение.. )
			System.out.println("Bad URL");
		} catch (IOException e) {  // ( ловим общее исключение.. )
			System.out.println("Problem in reading data: " + e.getMessage());
		}
	}
}

File: src/main/scala/com/learning/scala/controlstruct/ExceptionSyntax.scala
package com.learning.scala.controlstruct

import java.net.URL
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.MalformedURLException
import java.io.IOException

/*
 * Scala: Исключения (Exception)
 * Все исключения в Scala являются RuntimeException, потому что Scala проверяет исключение только в RuntimeException. Поэтому во время компиляции исключения НЕ обрабатываются..
 */


object ExceptionSyntax {

  def main(args: Array[String]) {
    val url = new URL("http://tirthalpatel.blogspot.com")

    /* try-catch-finally */
    try {
      val url = new URL( "http://tirthalpatel.blogspot.com" );
      val buffered = new BufferedReader( new InputStreamReader(url.openStream()) )
      try {
        var line = buffered.readLine
        while(line != null) {
          line = buffered.readLine
          println(line)
        }
      } finally {
        buffered.close()
      } 
    } catch {  // ( во время компиляции исключения НЕ будут проверены компилятором.. )
        case _: MalformedURLException => println("Bad URL")
        case e: IOException => println("Problem in reading data: " + e.getMessage)
    }
  }
}
